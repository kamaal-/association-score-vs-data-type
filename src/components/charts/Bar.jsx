import React, {useRef, useEffect} from 'react'
import { scaleLinear,  select, scaleBand, axisLeft, axisBottom} from 'd3'
import {setBars} from './helper'
import { COLORS } from '../../config'

const margin = {
    top: 20,
    left: 60,
    right: 50,
    bottom: 80
}

function Bar({row, title}) {
    let svg = null
    const svgRef = useRef()
    const wrapperRef = useRef()
    const data = row.visualizerData
    const headerHeight = 40
    useEffect(() => {
        const $el = svgRef.current
        const $wrapper = wrapperRef.current
        if($el && $wrapper){
            const height = $wrapper.offsetHeight
            const width = $wrapper.offsetWidth
            const computedHeight = height - (margin.top + margin.bottom + headerHeight)
            const computedWidth = width - (margin.left + margin.right)
            // eslint-disable-next-line react-hooks/exhaustive-deps
            svg = select($el)

            const svgWrapper = svg.append('g').attr('transform', `translate(${margin.left}, ${margin.top})`)

            const y = scaleLinear()
                .range([computedHeight, 0])
                .domain([0, 1])
            const x = scaleBand()
                            .range([0, computedWidth])
                            .domain(data.map(d => d.label))
                            .padding(0.2);

            svg.append('defs')
                .append('clipPath')
                .attr('id', 'clip')
                .attr('data-testid', 'clip')
                .append('rect')
                    .attr('x', 0)
                    .attr('y', 0)
                    .attr('width', computedWidth)
                    .attr('height', computedHeight);

            const plot = svgWrapper
                .append('g')
                .attr('class', 'plot')
                .attr('clip-path', 'url(#clip)');    
            
            plot.append('g')
                .attr('class', 'horizontal__grids')
                .attr('data-testid', 'h-grid')
                .call(
                    axisLeft(y)
                    .ticks(10)
                    .tickFormat("")
                    .tickSize(-computedWidth)
                );

            setBars(plot, x.bandwidth()/2, computedHeight, x, y, data,'value')
            
            svgWrapper.append('g')
                .attr('class', 'vertical__axis')
                .call(
                    axisLeft(y)
                    .ticks(10)
                );

            const xAxis = svgWrapper.append("g")
                .attr('class', 'horizontal__axis')
                .attr('transform', `translate(0,${computedHeight})`)
                .call(
                    axisBottom(x)
                );
            
            xAxis.selectAll("text")
                .attr("y", 0)
                .attr("dx", -10)
                .attr("transform", "rotate(-45)").style("text-anchor", "end");
        

            // Y label
            svgWrapper
                .append('g')
                .attr('transform', 'rotate(-90)')
                .append('text')
                .attr('data-testid', 'y-label')
                .attr('x', -computedHeight/2)
                .attr('y', -40)
                .style('text-anchor', 'middle')
                .text('Score')
                .attr('fill', COLORS.primaryTextColor)

            // X label
            svgWrapper
                .append('g')
                .append('text')
                .attr('data-testid', 'x-label')
                .attr('x', (computedWidth/2) - 20)
                .attr('y', computedHeight + 90)
                .style('text-anchor', 'middle')
                .text('Data Type')
                .attr('fill', COLORS.primaryTextColor)
        }
    })
    return (
        <div style={{height: '100%'}} ref={wrapperRef}>
            <h4 style={{margin: 0, height: '10px', marginBottom: '10px'}}>{title}</h4>
            <svg style={{width: '100%', height: `calc(100% - 20px)`}} ref={svgRef}/>
        </div>
    )
}

export default Bar
