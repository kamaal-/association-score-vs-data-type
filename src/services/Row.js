import {initialDataTypes} from '../config/'

export default class Row {
    constructor(data){
        this.data = data 
        this.setupDataTypes()
    }

    setupDataTypes(data){
        const self = this
        this.dataTypeKeyValues = (Object.keys(initialDataTypes)|| []).map( key => {
            return {...initialDataTypes[key], value: self.data.association_score.datatypes[key]}
        })
    }

    get geneId(){
        return this.data.target.id
    }

    get symbol(){
        return this.data.target.gene_info.symbol
    }

    get name(){
        return this.data.target.gene_info.name
    }

    get overallScore(){
        return this.data.association_score.overall
    }

    get visualizerData(){
        return this.dataTypeKeyValues
    }

    set dataType(dataType){
        this.dataTypeKeyValues.push(dataType)
    }
}