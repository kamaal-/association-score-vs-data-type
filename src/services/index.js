import Axios from "axios"
import Row from './Row'
import {totalItemsFromTo} from '../config/'

export const dataToRow = d => new Row(d)

export const transformDataToRows  = data => ((data && data.data) || [])
    .map(dataToRow)
    .sort(sortRowByOverallScore)
    .slice(...totalItemsFromTo)

export const sortRowByOverallScore = (a, b) => {
    if(a.overallScore > b.overallScore) return -1;
    if(a.overallScore < b.overallScore) return 1;
    return 0;
}

export const getInitialData = (cb) => {
    Axios.get('https://demo6922545.mockable.io/')
        .then(({data}) => cb(transformDataToRows (data)))
        .catch(e => {
            console.log(e)
        })
}