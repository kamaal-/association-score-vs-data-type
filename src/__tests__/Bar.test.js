import React from 'react';
import { render } from '@testing-library/react';
import Bar from '../components/charts/Bar';
import data from '../config/mock.data';
import { transformDataToRows Â } from '../services/'
import ReactDOM from 'react-dom'

describe('Bar Chart', () => {

    const setup = () => {
        const container = document.createElement('div')
        const rows = transformDataToRows(data())	
        return render(<Bar row={rows[0]} chartTitle={'Chart Title'}/>, {
            container: document.body.appendChild(container)
        });
    }

  	it('renders without crashing', () => {
        const div = document.createElement('div');
        const rows = transformDataToRows(data())	
    	ReactDOM.render(<Bar row={rows[0]} chartTitle={'Chart Title'}/>, div);
    	ReactDOM.unmountComponentAtNode(div);
    });

  	it('Should contain clip', () => {
		const { getAllByTestId } = setup()
		expect( getAllByTestId('clip').length).toBe(1);
    });

    it('Should contain horizontal grid', () => {
		const { getAllByTestId } = setup()
		expect( getAllByTestId('h-grid').length).toBe(1);
    });

    it('Should contain Score label', () => {
		const { getAllByTestId } = setup()
		expect( getAllByTestId('y-label').length).toBe(1);
    });

    it('Should contain Data Type label', () => {
		const { getAllByTestId } = setup()
		expect( getAllByTestId('x-label').length).toBe(1);
    });

    it('Should contain 7 bars', () => {
		const { getAllByTestId } = setup()
		expect( getAllByTestId('bar').length).toBe(7);
    });

    it('should match snapshot', () => {
        const { asFragment } = setup()
        expect(asFragment()).toMatchSnapshot();
    })

})
 