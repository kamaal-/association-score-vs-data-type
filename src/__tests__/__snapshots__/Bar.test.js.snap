// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bar Chart should match snapshot 1`] = `
<DocumentFragment>
  <div
    style="height: 100%;"
  >
    <h4
      style="margin: 0px 0px 10px 0px; height: 10px;"
    />
    <svg
      style="width: 100%;"
    >
      <g
        transform="translate(60, 20)"
      >
        <g
          class="plot"
          clip-path="url(#clip)"
        >
          <g
            class="horizontal__grids"
            data-testid="h-grid"
            fill="none"
            font-family="sans-serif"
            font-size="10"
            text-anchor="end"
          >
            <path
              class="domain"
              d="M-110,-139.5H0.5V0.5H-110"
              stroke="currentColor"
            />
            <g
              class="tick"
              opacity="1"
              transform="translate(0,-139.5)"
            >
              <line
                stroke="currentColor"
                x2="-110"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-113"
              />
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,-125.5)"
            >
              <line
                stroke="currentColor"
                x2="-110"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-113"
              />
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,-111.5)"
            >
              <line
                stroke="currentColor"
                x2="-110"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-113"
              />
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,-97.5)"
            >
              <line
                stroke="currentColor"
                x2="-110"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-113"
              />
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,-83.5)"
            >
              <line
                stroke="currentColor"
                x2="-110"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-113"
              />
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,-69.5)"
            >
              <line
                stroke="currentColor"
                x2="-110"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-113"
              />
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,-55.5)"
            >
              <line
                stroke="currentColor"
                x2="-110"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-113"
              />
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,-41.50000000000001)"
            >
              <line
                stroke="currentColor"
                x2="-110"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-113"
              />
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,-27.499999999999993)"
            >
              <line
                stroke="currentColor"
                x2="-110"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-113"
              />
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,-13.499999999999996)"
            >
              <line
                stroke="currentColor"
                x2="-110"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-113"
              />
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,0.5)"
            >
              <line
                stroke="currentColor"
                x2="-110"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-113"
              />
            </g>
          </g>
          <rect
            class="area__plot"
            data-testid="bar"
            fill="#ec7d20"
            height="-45.689479253864704"
            width="12.222222222222221"
            x="-15.277777777777786"
            y="-94.3105207461353"
          />
          <rect
            class="area__plot"
            data-testid="bar"
            fill="#f4ac4a"
            height="-9.09412116723928"
            width="12.222222222222221"
            x="-30.555555555555557"
            y="-130.90587883276072"
          />
          <rect
            class="area__plot"
            data-testid="bar"
            fill="#f6d74a"
            height="-140"
            width="12.222222222222221"
            x="-45.833333333333336"
            y="0"
          />
          <rect
            class="area__plot"
            data-testid="bar"
            fill="#cedc7e"
            height="-140"
            width="12.222222222222221"
            x="-61.111111111111114"
            y="0"
          />
          <rect
            class="area__plot"
            data-testid="bar"
            fill="#83a45a"
            height="-140"
            width="12.222222222222221"
            x="-76.38888888888889"
            y="0"
          />
          <rect
            class="area__plot"
            data-testid="bar"
            fill="#16868b"
            height="0"
            width="12.222222222222221"
            x="-91.66666666666667"
            y="-140"
          />
          <rect
            class="area__plot"
            data-testid="bar"
            fill="#2c63a0"
            height="-140"
            width="12.222222222222221"
            x="-106.94444444444444"
            y="0"
          />
        </g>
        <g
          class="vertical__axis"
          fill="none"
          font-family="sans-serif"
          font-size="10"
          text-anchor="end"
        >
          <path
            class="domain"
            d="M-6,-139.5H0.5V0.5H-6"
            stroke="currentColor"
          />
          <g
            class="tick"
            opacity="1"
            transform="translate(0,-139.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              0.0
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,-125.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              0.1
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,-111.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              0.2
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,-97.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              0.3
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,-83.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              0.4
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,-69.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              0.5
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,-55.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              0.6
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,-41.50000000000001)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              0.7
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,-27.499999999999993)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              0.8
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,-13.499999999999996)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              0.9
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,0.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              1.0
            </text>
          </g>
        </g>
        <g
          class="horizontal__axis"
          fill="none"
          font-family="sans-serif"
          font-size="10"
          text-anchor="middle"
          transform="translate(0,-140)"
        >
          <path
            class="domain"
            d="M0.5,6V0.5H-109.5V6"
            stroke="currentColor"
          />
          <g
            class="tick"
            opacity="1"
            transform="translate(-9.166666666666675,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dx="-10"
              dy="0.71em"
              fill="currentColor"
              style="text-anchor: end;"
              transform="rotate(-45)"
              y="0"
            >
              Literature
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(-24.444444444444446,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dx="-10"
              dy="0.71em"
              fill="currentColor"
              style="text-anchor: end;"
              transform="rotate(-45)"
              y="0"
            >
              RNA Expression
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(-39.72222222222223,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dx="-10"
              dy="0.71em"
              fill="currentColor"
              style="text-anchor: end;"
              transform="rotate(-45)"
              y="0"
            >
              Genetic Association
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(-55,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dx="-10"
              dy="0.71em"
              fill="currentColor"
              style="text-anchor: end;"
              transform="rotate(-45)"
              y="0"
            >
              Somatic Mutation
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(-70.27777777777777,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dx="-10"
              dy="0.71em"
              fill="currentColor"
              style="text-anchor: end;"
              transform="rotate(-45)"
              y="0"
            >
              Known Drug
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(-85.55555555555556,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dx="-10"
              dy="0.71em"
              fill="currentColor"
              style="text-anchor: end;"
              transform="rotate(-45)"
              y="0"
            >
              Animal Model
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(-100.83333333333333,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dx="-10"
              dy="0.71em"
              fill="currentColor"
              style="text-anchor: end;"
              transform="rotate(-45)"
              y="0"
            >
              Affected Pathway
            </text>
          </g>
        </g>
        <g
          transform="rotate(-90)"
        >
          <text
            data-testid="y-label"
            fill="#412960"
            style="text-anchor: middle;"
            x="70"
            y="-40"
          >
            Score
          </text>
        </g>
        <g>
          <text
            data-testid="x-label"
            fill="#412960"
            style="text-anchor: middle;"
            x="-75"
            y="-50"
          >
            Data Type
          </text>
        </g>
      </g>
      <defs>
        <clippath
          data-testid="clip"
          id="clip"
        >
          <rect
            height="-140"
            width="-110"
            x="0"
            y="0"
          />
        </clippath>
      </defs>
    </svg>
  </div>
</DocumentFragment>
`;
